frontend/src/calldata/encoder.ts

```diff
@@ -0,0 +1,161 @@
+import type { CalldataEncodable } from './types';
+import { Address } from './types';
+
+const BITS_IN_TYPE = 3;
+
+const TYPE_SPECIAL = 0;
+const TYPE_PINT = 1;
+const TYPE_NINT = 2;
+const TYPE_BYTES = 3;
+const TYPE_STR = 4;
+const TYPE_ARR = 5;
+const TYPE_MAP = 6;
+
+const SPECIAL_NULL = (0 << BITS_IN_TYPE) | TYPE_SPECIAL;
+const SPECIAL_FALSE = (1 << BITS_IN_TYPE) | TYPE_SPECIAL;
+const SPECIAL_TRUE = (2 << BITS_IN_TYPE) | TYPE_SPECIAL;
+const SPECIAL_ADDR = (3 << BITS_IN_TYPE) | TYPE_SPECIAL;
+
+function reportError(msg: string, data: CalldataEncodable): never {
+  throw new Error(`invalid calldata input '${data}'`);
+}
+
+function writeNum(to: number[], data: bigint) {
+  if (data === 0n) {
+    to.push(0);
+    return;
+  }
+  while (data > 0) {
+    let cur = Number(data & 0x7fn);
+    data >>= 7n;
+    if (data > 0) {
+      cur |= 0x80;
+    }
+    to.push(cur);
+  }
+}
+
+function encodeNumWithType(to: number[], data: bigint, type: number) {
+  const res = (data << BigInt(BITS_IN_TYPE)) | BigInt(type);
+  writeNum(to, res);
+}
+
+function encodeNum(to: number[], data: bigint) {
+  if (data >= 0n) {
+    encodeNumWithType(to, data, TYPE_PINT);
+  } else {
+    encodeNumWithType(to, -data - 1n, TYPE_NINT);
+  }
+}
+
+function compareString(l: Uint8Array, r: Uint8Array): number {
+  // FIXME it is incorrect for non ascii!
+  for (let index = 0; index < l.length && index < r.length; index++) {
+    const cur = l[index] - r[index];
+    if (cur !== 0) {
+      return cur;
+    }
+  }
+  return l.length - r.length;
+}
+
+function encodeMap(to: number[], arr: Iterable<[string, CalldataEncodable]>) {
+  const newEntries = Array.from(
+    arr,
+    ([k, v]): [Uint8Array, CalldataEncodable] => [
+      new TextEncoder().encode(k),
+      v,
+    ],
+  );
+  newEntries.sort(([k1, _v1], [k2, _v2]) => compareString(k1, k2));
```

Comment:
eslint gives a warning about `_v1` and `_v2` being unused even though they are prefixed with an underscore. How to silence that?...

Reply:
@kp2pml30 they are not being used, when calling compareString(k1, k2) so you can totally omit them.

Reply:
Okay, js pattern matching is as surprising as always